### ISSUE TITLE
Lab3 Peer Review by Federico Boscolo (s294908)

### ISSUE THREAD

**Issue by federicoboscolo** [Leo-Rola/Computational-Intelligence-2022-2023#6]
Hi Leonardo, since you reviewed my lab I chose you for my review. :)
I am running the code on an i5 Windows machine.

**Task 3.1**
First of all, I want to start by saying that making the game interactive by default is a very nice idea, I actually had a bit of fun playing against your fixed strategy and it really does feel like a smart agent! 

Your idea of seeing one move ahead by having it play against itself is very smart and I hadn’t thought about it. However, I must say that I had some difficulty understanding the logic behind the code, and I believe it could benefit from a little refactoring and maybe a further explanation on how the Evaluator operates. Also, I was able to look at the source of the (marginal) slowdowns while playing against it, which is your 100,000 iterations limit on the infinite loop caused by a winning state for the opponent. I don’t believe it is necessary to make it run that long since in other cases it’s able to find a move much more quickly. Maybe you could exchange it for a check on the state instead of trying to find the optimal move?
These two observations aside, the code is effective and covers a wide range of possibilities, and you actually managed to get close to nim sum so again, very good job overall!

**Task 3.2**
The evolvable strategy uses a move-by-move evolutionary approach, which I think is a great idea and quite effective as well. The idea of the crossover function averaging the probability thresholds of the parents is smart and does a good job of increasing the average fitness with each generation. In the end, the strategy pretty much converges to the max_lowest strategy, and it averages an 80% winrate against the random strategy. Good work! I found the code very clear to read as well.
Just as an aside on the readme, I think a table with the results against random, optimal and fixed strategy could’ve been handy. To get the winrate I mentioned (80%) I ran the evaluate function manually using the highest fitness individual.

**Task 3.3**
I’ll start off by saying great job on the pruning, getting results for N=6 in under two minutes is a very good achievement considering the computational overhead that minmax requires. The code is clear and actually has quite a simple structure that makes it very readable.
I am curious as to why you implemented a terminal state evaluation of (-1, 0) instead of (-1, +1) as is found in most documentation about minmax. The code obviously works great with it, it’s just that I haven’t seen it anywhere. Great results overall, with a 100% winrate on random and fixed strategy.
One small optimization I’d suggest would be to store the score of a given state once calculated, I believe it would reduce the computation time quite significantly.
Once again, very very good work!

**Task 3.4**
The reinforcement learning agent shows very good results, surpassing all other strategies to win as much as 85% of the time against the optimal strategy. Impressive! Very smart move to train it against the random strategy as well as the optimal one, and it does pay off.
The code itself is very straightforward and easy to read, as it shares the same structure of the maze exercise which we’re all familiar with. The implementation is proven to be effective in this case.
The 0% winrate with higher game sizes might just be based on who starts first, because the optimal strategy is heavily dependent on the game size and turn. I would try training the agent both starting first and second, and evaluate it as such too. I am confident it can bring you great results with the code you have written.

In conclusion, the lab is very well done and it achieves great results overall, so I think that speaks for itself. Thanks again for reviewing my code and have a great night!
